File: java-to-kotlin/src/main/java/nl/rabobank/kotlinmovement/recipes/service/RecipesService.java
package nl.rabobank.kotlinmovement.recipes.service;

import lombok.AllArgsConstructor;
import nl.rabobank.kotlinmovement.recipes.data.IngredientsEntity;
import nl.rabobank.kotlinmovement.recipes.data.IngredientsRepository;
import nl.rabobank.kotlinmovement.recipes.data.RecipesEntity;
import nl.rabobank.kotlinmovement.recipes.data.RecipesRepository;
import nl.rabobank.kotlinmovement.recipes.model.RecipeRequest;
import nl.rabobank.kotlinmovement.recipes.model.RecipeResponse;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import static nl.rabobank.kotlinmovement.recipes.service.RecipesMapper.*;

@Service
@AllArgsConstructor
public class RecipesService {
    @NotNull
    private final RecipesRepository recipeRepository;
    @NotNull
    private final IngredientsRepository ingredientsRepository;

    @Transactional
    public RecipeResponse getRecipe(long id) {
        var recipe = recipeRepository.findById(id);
        return recipe.map(r -> toRecipeResponse(r, r.getIngredients()))
                .orElseThrow(() -> new ResourceNotFoundException("Recipe %d not found".formatted(id)));
    }

    @NotNull
    @Transactional
    public List<RecipeResponse> getRecipes() {
        return recipeRepository.findAll()
                .stream()
                .map(r -> toRecipeResponse(r, r.getIngredients()))
                .collect(Collectors.toList());
    }

    @NotNull
    @Transactional
    public RecipeResponse saveRecipe(RecipeRequest recipeRequest) {
        final RecipesEntity recipe = toRecipeEntity(recipeRequest);
        var recipes = recipeRepository.save(recipe);
        var ingredients = saveIngredients(recipeRequest, recipes);
        return toRecipeResponse(recipes, ingredients);
    }

    @NotNull
    @Transactional
    public RecipeResponse updateOrCreateRecipe(Long id, RecipeRequest recipeRequest) {
        return recipeRepository.findById(id)
                .map(it -> {
                    var recipe = recipeRepository.save(new RecipesEntity(it.getId(), recipeRequest.getRecipeName(), Collections.emptySet()));
                    var ingredients = saveIngredients(recipeRequest, recipe);
                    return toRecipeResponse(recipe, ingredients);
                }).
                orElse(saveRecipe(recipeRequest));
    }

    @Transactional
    public void deleteRecipe(long id) {
            recipeRepository.deleteById(id);
    }

    @NotNull
    private Set<IngredientsEntity> saveIngredients(RecipeRequest recipeRequest, RecipesEntity recipe) {
        final Set<IngredientsEntity> ingredients = toIngredientsEntity(recipeRequest, recipe);
        return new HashSet<>(ingredientsRepository.saveAll(ingredients));
    }

}

File: java-to-kotlin-complete/src/main/kotlin/nl/rabobank/kotlinmovement/recipes/service/RecipesService.kt
package nl.rabobank.kotlinmovement.recipes.service

import nl.rabobank.kotlinmovement.recipes.data.IngredientsEntity
import nl.rabobank.kotlinmovement.recipes.data.IngredientsRepository
import nl.rabobank.kotlinmovement.recipes.data.RecipesEntity
import nl.rabobank.kotlinmovement.recipes.data.RecipesRepository
import nl.rabobank.kotlinmovement.recipes.model.RecipeRequest
import nl.rabobank.kotlinmovement.recipes.model.RecipeResponse
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class RecipesService(
    val recipeRepository: RecipesRepository,
    val ingredientsRepository: IngredientsRepository
) {

    @get:Transactional
    val recipes: List<RecipeResponse>
        get() = recipeRepository.findAll()
            .mapNotNull { r: RecipesEntity -> r.ingredients?.let { RecipesMapper.toRecipeResponse(r, r.ingredients) } }

    @Transactional
    fun getRecipe(id: Long): RecipeResponse? = recipeRepository.findByIdOrNull(id)
        ?.let { recipesEntity ->
            checkNotNull(recipesEntity.ingredients)
            RecipesMapper.toRecipeResponse(recipesEntity, recipesEntity.ingredients)
        }
        ?: throw ResourceNotFoundException("Recipe $id not found")

    @Transactional
    fun saveRecipe(recipeRequest: RecipeRequest): RecipeResponse? {
        val recipe = RecipesMapper.toRecipeEntity(recipeRequest)
        val recipes = recipeRepository.save(recipe)
        val ingredients = saveIngredients(recipeRequest, recipes)
        return RecipesMapper.toRecipeResponse(recipes, ingredients)
    }

    @Transactional
    fun updateOrCreateRecipe(id: Long, recipeRequest: RecipeRequest): RecipeResponse? {
        return recipeRepository.findByIdOrNull(id)?.let {
            requireNotNull(recipeRequest.recipeName)
            val recipe = recipeRepository.save(RecipesEntity(it.recipeId, recipeRequest.recipeName))
            val ingredients = saveIngredients(recipeRequest, recipe)
            return RecipesMapper.toRecipeResponse(recipe, ingredients)
        } ?: saveRecipe(recipeRequest)
    }

    @Transactional
    fun deleteRecipe(id: Long) {
            recipeRepository.deleteById(id)
    }

    private fun saveIngredients(recipeRequest: RecipeRequest, recipe: RecipesEntity): Set<IngredientsEntity> =
        ingredientsRepository.saveAll(RecipesMapper.toIngredientsEntity(recipeRequest, recipe)).toSet()
}
